lib/HLSL/DxilValidationRaceCondCheck.cpp

```diff
@@ -0,0 +1,85 @@
+//===-- DxilValidationRaceCondCheck.cpp - Dxil race condition check pass ---===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// \file
+// Check race condition for Dxil Validation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "DxilTargetTransformInfo.h"
+#include "dxc/HLSL/DxilModule.h"
+#include "dxc/HLSL/DxilOperations.h"
+#include "DxilValidationRaceCondCheck.h"
+
+#include "llvm/Analysis/PostDominators.h"
+#include "llvm/CodeGen/BasicTTIImpl.h"
+#include "llvm/Analysis/DivergenceAnalysis.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instructions.h"
+
+
+using namespace llvm;
+using namespace hlsl;
+
+namespace llvm {
+void initializeRaceConditionCheckPassPass(PassRegistry&);
+
+Pass *createRaceConditionCheckPass(std::vector<StoreInst *> *TGSMList,
+                                   std::unordered_set<Function *> *FuncSet,
+                                   RaceConditionCheckPass::ErrorEmitter *EE) {
```

Comment:
Why pass in the instructions and function lists? Seems like it would make more sense to generate these inside the runOnFunction() in the pass. In fact the function list would not even be needed if you handle everything inside runOnFunction()